costs/include/finite_diff_cost.h

```diff
@@ -0,0 +1,93 @@
+//
+// Created by gavin on 4/26/2024.
+//
+
+#ifndef TORC_FINITE_DIFF_COST_H
+#define TORC_FINITE_DIFF_COST_H
+
+#include "base_cost.h"
+#include <iostream>
+
+namespace torc {
+    /**
+     * Class implementation of a cost function whose differentials are evaluated using the finite difference method.
+     * @tparam scalar_t the type of scalar used for the cost
+     */
+    template <class scalar_t>
+    class FiniteDiffCost: public BaseCost<scalar_t> {
+        using vectorx_t = Eigen::VectorX<scalar_t>;
+        using matrixx_t = Eigen::MatrixX<scalar_t>;
+    public:
+        /**
+         * Constructor for the Finite Difference Cost class
+         * @param fn       the cost function
+         * @param dim           the dimensions of the input of the cost function
+         * @param grad_step     the step used for the gradient
+         * @param hess_step     the step used for the hessian
+         * @param identifier    a string identifier of the cost function
+         */
+        FiniteDiffCost(const std::function<scalar_t(vectorx_t)>& fn,
+                       const size_t& dim,
+                       const scalar_t grad_step=3e-8,   // approximately sqrt(ulp)
+                       const scalar_t hess_step=1e-3,   // approximately sqrt(grad_step)
+                       const std::string& identifier="Finite Difference Cost Instance") {
+            this->identifier_ = identifier;
+            this->fn_ = fn;
+            this->dim_ = dim;
+            this->grad_step_ = grad_step;
+            this->hess_step_ = hess_step;
+        }
+
+        /**
+         * Evaluates the cost function at a given point
+         * @param x the input to the function
+         * @return f(x)
+         */
+        scalar_t Evaluate(const vectorx_t& x) const {
+            return fn_(x);
+        }
+
+        /**
+         * Returns the gradient of the cost evaluated at x using the central finite difference method
+         * @param x the input
+         * @return grad f(x)
+         */
+        vectorx_t Gradient(const vectorx_t& x) const {
+            unsigned dim = this->dim_;
```

Comment:
If you are going to assign the member variable to a local variable (I imagine so the code is less cluttered), it would be good to make this `const` since the function is a const member function. This way we don't accidentally end up modifying the local copy of what should be a const variable. This may also let the compiler optimize this assignment out entirely (honestly modern compilers might be able to that anyway as they can see you never modify it, but it doesn't hurt to add the `const`).

As a general note, in your code any variable that doesn't need to be modified should be const. So in this function you can make `dim` and `pos_diff` and `neg_diff` const. Not a huge deal but I think it's best practice.